tests/unit_and_functional/test_workflow_execution.py

```diff
@@ -161,3 +166,46 @@ def test_run_workflow_adds_git_commit_to_batch_and_service_env_vars(
         cpu_request=ANY,
         memory_request=ANY,
     )
+
+
+@patch("bodywork.workflow_execution.rmtree")
+@patch("bodywork.workflow_execution.requests.Session")
+@patch("bodywork.workflow_execution.download_project_code_from_repo")
+@patch("bodywork.workflow_execution.get_git_commit_hash")
+@patch("bodywork.workflow_execution.k8s")
+def test_run_workflow_pings_usage_stats_server(
+    mock_k8s: MagicMock,
+    mock_git_hash: MagicMock,
+    mock_git_download: MagicMock,
+    mock_session: MagicMock,
+    mock_rmtree: MagicMock,
+    project_repo_location: Path,
+):
+    config_path = Path(f"{project_repo_location}/bodywork.yaml")
+    config = BodyworkConfig(config_path)
+    config.project.usage_stats = True
+
+    run_workflow(config, "foo_bar_foo_993", project_repo_location)
+
+    mock_session().get.assert_called_with(USAGE_STATS_SERVER_URL)
+
+
+@patch("bodywork.workflow_execution.rmtree")
+@patch("bodywork.workflow_execution.requests.Session")
+@patch("bodywork.workflow_execution.download_project_code_from_repo")
+@patch("bodywork.workflow_execution.get_git_commit_hash")
+@patch("bodywork.workflow_execution.k8s")
+def test_usage_stats_opt_out_does_not_ping_usage_stats_server(
+    mock_k8s: MagicMock,
+    mock_git_hash: MagicMock,
+    mock_git_download: MagicMock,
+    mock_session: MagicMock,
+    mock_rmtree: MagicMock,
+    project_repo_location: Path,
+):
+    config_path = Path(f"{project_repo_location}/bodywork.yaml")
+    config = BodyworkConfig(config_path)
+
+    run_workflow(config, "foo_bar_foo_993", project_repo_location)
+
+    mock_session().get.assert_called_once()
```

Comment:
Shouldn't it be checking that it hasn't been called?

Reply:
It gets called once by the docker image request so we're checking that it's only called once. Unfortunately there isn't a `assert_not_called_with` in pytest.

Reply:
`not mock_session().get.assert_called_with(USAGE_STATS_SERVER_URL)`?

Reply:
If only it was that easy! The assert is in the mock method call so the test will finish with that `assert_called_with()`.

Reply:
🤔

```python
try:
    mock_session().get.assert_called_with(USAGE_STATS_SERVER_URL)
    assert False
except Exception:
    assert True
```

?

Reply:
That looks very hacky. I think it will always return true if a false assertion always creates an exception.

Reply:
I agree...